morpheus/_lib/src/stages/inference_client_stage.cpp

```diff
@@ -0,0 +1,276 @@
+{
+    // When our tensor lengths are longer than our dataframe we will need to use the seq_ids array to
+    // lookup how the values should map back into the dataframe.
+    auto host_seq_ids = get_seq_ids(x);
+
+    for (auto& mapping : output_tensors)
+    {
+        auto& output_tensor = mapping.second;
+
+        morpheus::ShapeType shape  = output_tensor.get_shape();
+        morpheus::ShapeType stride = output_tensor.get_stride();
+
+        morpheus::ShapeType reduced_shape{shape};
+        reduced_shape[0] = x->mess_count;
+
+        auto reduced_buffer = morpheus::MatxUtil::reduce_max(
+            morpheus::DevMemInfo{
+                output_tensor.data(), output_tensor.dtype(), output_tensor.get_memory(), shape, stride},
+            host_seq_ids,
+            0,
+            reduced_shape);
+
+        output_tensor.swap(
+            morpheus::Tensor::create(std::move(reduced_buffer), output_tensor.dtype(), reduced_shape, stride, 0));
+    }
+}
+
+static void apply_logits(morpheus::TensorMap& output_tensors)
+{
+    for (auto& mapping : output_tensors)
+    {
+        auto& output_tensor = mapping.second;
+
+        auto shape  = output_tensor.get_shape();
+        auto stride = output_tensor.get_stride();
+
+        auto output_buffer = morpheus::MatxUtil::logits(morpheus::DevMemInfo{
+            output_tensor.data(), output_tensor.dtype(), output_tensor.get_memory(), shape, stride});
+
+        // For logits the input and output shapes will be the same
+        output_tensor.swap(morpheus::Tensor::create(std::move(output_buffer), output_tensor.dtype(), shape, stride, 0));
+    }
+}
+
+}  // namespace
+
+namespace morpheus {
+
+InferenceClientStage::InferenceClientStage(std::unique_ptr<IInferenceClient>&& client,
+                                           std::string model_name,
+                                           bool needs_logits,
+                                           std::vector<TensorModelMapping> input_mapping,
+                                           std::vector<TensorModelMapping> output_mapping) :
+  m_model_name(std::move(model_name)),
+  m_client(std::move(client)),
+  m_needs_logits(needs_logits),
+  m_input_mapping(std::move(input_mapping)),
+  m_output_mapping(std::move(output_mapping))
+{}
+
+struct ExponentialBackoff
+{
+    std::shared_ptr<mrc::coroutines::Scheduler> m_on;
+    std::chrono::milliseconds m_delay;
+    std::chrono::milliseconds m_delay_max;
+
+    ExponentialBackoff(std::shared_ptr<mrc::coroutines::Scheduler> on,
+                       std::chrono::milliseconds delay_initial,
+                       std::chrono::milliseconds delay_max) :
+      m_on(std::move(on)),
+      m_delay(delay_initial),
+      m_delay_max(delay_max)
+    {}
+
+    mrc::coroutines::Task<> yield()
+    {
+        if (m_delay > m_delay_max)
+        {
+            m_delay = m_delay_max;
+        }
+
+        co_await m_on->yield_for(m_delay);
+
+        m_delay *= 2;
+    }
+};
```

Comment:
This would be useful as a class in a separate file so other stages could use it.